### ISSUE TITLE
Formatted text

### ISSUE THREAD

**Issue by georgefst** [diagrams/diagrams#21]
I've just converted a project from directly using the `graphviz` package to using `diagrams-graphviz`, and it's opened up a whole new world of options. But the one thing I'm missing is text formatting - my graphs have edges labelled by fairly complex logical expressions, which become much more readable with syntax highlighting.

After considering a few approaches, I've written a function for rendering a `Diagram` from a [SimpleDocStream](https://hackage.haskell.org/package/prettyprinter-1.6.1/docs/Data-Text-Prettyprint-Doc.html#t:SimpleDocStream) annotated with `Style`s, by splitting the text when formatting changes, and concatenating the resulting diagrams.

This approach is rather conceptually ugly, and only works well with monospace fonts. Seeing as I'm using the `cairo` backend, results would be nicer if I were able to more directly make use of `pango`. 

There seem two major options:
- Expose `pango`'s rendering functionality directly in `diagrams-cairo` with a function along the lines of `SomeOptionsType -> [PangoItem] -> Diagram Cairo`. `pango` is already a dependency, but much of its power seems to be unused.
- Add an annotated text type to `diagrams-core`, perhaps along the lines of `prettyprinter`'s `SimpleDocStream` (mentioned above). Many of the other backends could probably implement rendering in a sensible way.

**Comment by byorgey** [comment]
Hi @georgefst !  I definitely think this sounds feasible, and would be happy to take a PR.  It probably makes the most sense to start with a `diagrams-cairo`/`pango`-specific feature.  Once that works well we could think about adding something in `diagrams-core` and updating other backends (which would be much more work).

**Comment by georgefst** [comment]
I've looked into this in some detail, and it seems like the `pango` bindings don't expose the methods that would be needed to do this neatly. The newer `gi-pango` library, however, does...

Indeed, seeing as the auto-generated, `gi-` bindings seem to be the future for using GTK etc. through Haskell, I may look into making an experimental fork of `diagrams-cairo` to use those instead. That's more work than I initially had in mind, but there would be other benefits.

**Comment by byorgey** [comment]
Makes sense.  Let us know if we can answer any questions or help in any way!

**Comment by georgefst** [comment]
I notice that there is a `test` folder, but no test suite, so I assume regression testing consists of regenerating diagrams and reviewing manually (or indeed, comparing equality of the corresponding SVG files, though obviously there could be false positives)?

How comprehensive are the tests?

**Comment by byorgey** [comment]
I don't think there is much of a test suite in the diagrams-cairo repo
itself.  Take a look at https://github.com/diagrams/diagrams-backend-tests .
